uast/ann/ann.go

```diff
@@ -0,0 +1,219 @@
+// Package ann provides a DSL to annotate UAST.
+package ann
+
+import (
+	"github.com/bblfsh/sdk/uast"
+)
+
+// PathMatcher is a function that matches against a NodePath. If a prefix is not
+// matched, it returns that prefix. If there is no match at all, it returns the
+// input NodePath. If it matches the full NodePath, nil is returned.
```

Comment:
Clarified PathMatcher godoc.

PathMatcher is an interface that can be implemented by anything. PathPredicate is a convenient way of signaling that something is a predicate that takes path, and as such, it implements the PathMatcher interface.

It is also implemented by NodePredicate.

Implementing PathMatcher interface directly in the predicates allows a more fluent DSL.

For example:
```
On(HasInternalType("root")).Role(File)
	Rules(
		On(Any(), HasInternalType("import")).Roles(ImportDeclaration),
		On(Not(HasInternalType("import"))).Roles(TypeDeclaration),
	)
```

`On` works on any PathMatcher, while `Not` works on NodePredicate. Implementing the PathMatcher interface in NodePredicate allows combining them fluently.

Any type change is welcome, but keep in mind that this package is a DSL oriented to make writing UAST annotators in a declarative way easily.